napari/utils/events/dataclass.py

```diff
@@ -180,13 +182,39 @@ def set_with_events(self: C, name: str, value: Any) -> None:
 
     # if different we emit the event with new value
     after = getattr(self, name)
-    if isinstance(before, np.ndarray) or isinstance(after, np.ndarray):
-        different = np.any(before != after)
-    else:
-        different = before != after
-    if different:
+    if not compare(after, before):
```

Comment:
I'm just looking at some of the complexity introduced into this class to handle the case where event emission should be blocked when the input value is the same as the output value. We used this pattern a lot in the codebase before the evented_dataclass, which is why I think it was added, but maybe we should step back and consider dropping it, i.e. if you ever try and set then we will emit event, even if value hasn't changed.

There might be some places in the codebase now where we will want to add if statements like this outside the dataclass, or maybe we can clean up some code to prevent us setting things that shouldn't be set.

Curious what @tlambert03 @jni @Czaki think of this idea?

Reply:
> if you ever try and set then we will emit event, even if value hasn't changed.

I'm definitely open to that.  i agree it's getting hairier than I wanted for something going on every attribute of every class

Reply:
> I'm definitely open to that. i agree it's getting hairier than I wanted for something going on every attribute of every class

Ok i might look into what the consequences of dropping this entirely would be in a new PR. Would be good to understand before making a decision. If we drop then that keeps these extra dependencies out of the class, which could really one day live with our event system in it's own standalone `events` library (if we so wanted)

Reply:
As I understand I'm against this. In many places in codebases signals are used to synchronize state (I think that Camer ecosystem could be an example). So if a guardian is not provided then It could come in an infinity loop. 

Also experience from another libs is that set of the same value will not call callbacks/emit signals.

Reply:
> In many places in codebases signals are used to synchronize state (I think that Camer ecosystem could be an example). So if a guardian is not provided then It could come in an infinity loop.

We'd definitely have to fix these, but that might improve the overall codebase!

> Also experience from another libs is that set of the same value will not call callbacks/emit signals.

Ok that's good to know. Is that the case with Qt?

Curious what @jni thinks here

Reply:
> Ok that's good to know. Is that the case with Qt?

As I remember Qt native widgets des not emit a signal if you set the current value.

Reply:
> We'd definitely have to fix these, but that might improve the overall codebase!

First, you need to dramatically increase the test codebase to add many integration tests between components. Otherwise, there will be many regressions that are hard to debug. 

And we do not start talking about user custom widgets. 

I would rather create a custom `property` like decorator which will do this check.

Reply:
So this is not a part of this PR and should be done in another PR?